plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitLoader.java

```diff
@@ -13,64 +13,140 @@
+        this.dynamicFilteringWaitTimeoutMillis = dynamicFilteringWaitTimeoutMillis;
+        this.timeZoneKey = requireNonNull(timeZoneKey, "timeZoneKey is null");
+        this.stopwatch = Stopwatch.createStarted();
     }
 
     @Override
     public TaskStatus process()
     {
+        long timeLeft = dynamicFilteringWaitTimeoutMillis - stopwatch.elapsed(MILLISECONDS);
+        if (dynamicFilter.isAwaitable() && timeLeft > 0) {
+            return TaskStatus.continueOn(asVoid(toListenableFuture(dynamicFilter.isBlocked().orTimeout(timeLeft, MILLISECONDS))));
+        }
+
         if (combinedScanIterable == null) {
-            this.combinedScanIterable = tableScan.planTasks();
+            Expression filterExpression = toIcebergExpression(
+                    tableHandle.getEnforcedPredicate()
+                            // TODO: Remove TupleDomain#simplify once Iceberg supports IN expression. Currently this
+                            // is required for IN predicates on non-partition columns with large value list. Such
+                            // predicates on partition columns are not supported.
+                            // (See AbstractTestIcebergSmoke#testLargeInFailureOnPartitionedColumns)
+                            .intersect(
+                                    tableHandle.getUnenforcedPredicate()
+                                            .intersect(dynamicFilter.getCurrentPredicate().transformKeys(IcebergColumnHandle.class::cast))
+                                            .simplify(ICEBERG_DOMAIN_COMPACTION_THRESHOLD)));
+            this.combinedScanIterable = tableScan
+                    .filter(filterExpression)
+                    .includeColumnStats()
+                    .planTasks();
             this.fileScanIterator = Streams.stream(combinedScanIterable)
                     .map(CombinedScanTask::files)
                     .flatMap(Collection::stream)
                     .iterator();
         }
 
+        Map<Integer, Type.PrimitiveType> primitiveTypeForFieldId = primitiveFieldTypes(tableScan.schema());
         while (fileScanIterator.hasNext()) {
             FileScanTask task = fileScanIterator.next();
             if (!task.deletes().isEmpty()) {
-                throw new TrinoException(NOT_SUPPORTED, "Iceberg tables with delete files are not supported: " + schemaTableName);
+                throw new TrinoException(NOT_SUPPORTED, "Iceberg tables with delete files are not supported: " + tableHandle.getSchemaTableName());
             }
-            ListenableFuture<Void> queueFullFuture = splitQueue.offer(toIcebergSplit(task));
+            IcebergSplit icebergSplit = toIcebergSplit(task);
+            TupleDomain<IcebergColumnHandle> dynamicFilterPredicate = dynamicFilter.getCurrentPredicate()
+                    .transformKeys(IcebergColumnHandle.class::cast);
+
+            if (!dynamicFilterPredicate.isAll()) {
+                if (!fileMatchesPredicate(
```

Comment:
We should evaluate the less granular filter first for the sake of efficiency, that seems to be the partition level filter. But we can keep the current evaluation order if you know that this predicate is cheaper to evaluate.

Reply:
Makes sense, switched